### PR TITLE
Remove py_func for cross-environment serializability. Convert Detecti…

### PR DESCRIPTION
**Pull request by Cpruce**
…onLayer/refine_detections to tf

### CODE REVIEW

**Comment by ps48** [comment]
@Cpruce  did you try running it end to end? I somehow got errors in adding this PR

**Comment by Cpruce** [comment]
@ps48 Thanks for trying it out. I wasn't sure when to create pull request. In short, Im having trouble converting one section to tensorflow. I'm looking for advice on how to translate:

`for class_id in np.unique(pre_nms_class_ids):
         # Pick detections of this class
         ixs = np.where(pre_nms_class_ids == class_id)[0]
         # Apply NMS
         class_keep = utils.non_max_suppression(
             pre_nms_rois[ixs], pre_nms_scores[ixs],
             config.DETECTION_NMS_THRESHOLD)
         # Map indicies
         class_keep = keep[ixs[class_keep]]
         nms_keep = np.union1d(nms_keep, class_keep)
     keep = np.intersect1d(keep, nms_keep).astype(np.int32)`

Currently, I have

`for class_id in np.unique(pre_nms_class_ids):
          #print('class_id: ', tf.unstack(class_id))
          # Pick detections of this class
          ixs = tf.where(pre_nms_class_ids == class_id)[0]
          # Apply NMS
          #class_keep = utils.non_max_suppression(
          #    pre_nms_rois[ixs], pre_nms_scores[ixs],
          #    config.DETECTION_NMS_THRESHOLD)
          class_keep = tf.image.non_max_suppression(
                  tf.to_float(tf.gather_nd(pre_nms_rois,ixs)),
                  tf.gather(pre_nms_scores, ixs),
                  ixs.shape[0],
                  iou_threshold=config.DETECTION_NMS_THRESHOLD)

          # Map indicies
          class_keep = tf.gather(keep, tf.gather(ixs, class_keep))
          #keep[ixs[class_keep]]
          if nms_keep == []:
              nms_keep = class_keep
          else:
              nms_keep = tf.unique(tf.concat([nms_keep, class_keep], 0))[0]
              #tf.sets.set_union(nms_keep, class_keep)
....
keep = tf.sets.set_intersection(keep, nms_keep)`

Yet the `set_intersection` function keeps failing on the input I give. I will continue debugging tonight. Just hoping someone would know quickly a better solution.

**Comment by ps48** [comment]
@Cpruce this part tricky, let's try to solve it together. 

1. The **for loop** needs to be replaced with a **tf.map_fn or tf.while_loop** so that we can iterate over indivudual elements, I would prefer while_loop as it gives more flexibility for TensorArray operations.

2. **tf.sets.set_intersection** needs atleast least rank 2 for Dense To Dense SetOperation. Hence a work around would be to concatenate keep & nms_keep, then apply unique_with_count operation to see which all elements are double in resulted concatenation. We shall retain the elements which have count atleast two or greater. 
**TL;DR** replace the set_intersection op with
`result_keep = tf.concat([keep,nms_keep], axis = 0)`
`output_keep, idx_keep, count_keep = tf.unique_with_counts(result_keep)`
`new_idx_keep = tf.where(count_keep >= tf.constant(2))`
`keep = tf.gather(output_keep, new_idx_keep)`

The work around is crude but should get the code going. @waleedka your advice on this would be of great help.

**Comment by Cpruce** [comment]
Awesome thanks @ps48 ! I'll try your suggestions tonight :)

**Comment by Cpruce** [comment]
@ps48 @waleedka I tried the while_loop approach and I think that I more-or-less have it. However, I am running into a problem with the '?' Dimension from `tf.unique`. It seems that the tensor dimension is too generic:

`ValueError: Cannot convert an unknown Dimension to a Tensor: ?`

I got map_fn to work. However, top_k hits the same problem in:

```python
roi_count = config.DETECTION_MAX_INSTANCES
class_scores_keep = tf.gather(class_scores, keep)
num_keep = min(max(class_scores_keep.get_shape()[0], 0), roi_count)
top_ids = tf.nn.top_k(class_scores_keep, k=num_keep, sorted=True)#[::-1]
```

```ValueError: Tried to convert 'k' to a tensor and failed. Error: Cannot convert an unknown Dimension to a Tensor: ?```

Should I be using a placeholder here? If not, got any ideas on workarounds? Thanks again!

**Comment by ps48** [comment]
@Cpruce nice to see the progress.

For tf.unique try reshaping the input to [-1] shape. 
`keep = tf.unique(tf.reshape(keep,[-1]))[0]`

I still feel **tf.map_fn** may give error for elements with multiple class ids. Try using **tf.while_loop** if possible because it has flexibility in stacking of result.

For the tf.nn.top_k you may use the following code. Convert the **ROI count to a tensor** from config.py and then consume it in **tf.minimum**. We cannot use min/max as it is a python function. Finally in **tf.nn.top_k** the returned elements **values and indices**, what we need is indices hence we use the second element.

`roi_count = tf.convert_to_tensor(config.DETECTION_MAX_INSTANCES)`
`class_scores_keep = tf.gather(class_scores, keep)`
`num_keep = tf.minimum(tf.shape(class_scores_keep)[0], roi_count)`
`top_ids = tf.nn.top_k(class_scores_keep, k=num_keep, sorted=True)[1]`

**Comment by Cpruce** [comment]
@ps48 Awesome thanks again! I've updated the branch and am almost able to make it to the end using map_fn. I will try while_loop again tonight. 

For padding,

If `gap` is a tensor, how should I make the tensor `paddings = tf.constant([[0, gap], [0, 0]])`?

Thanks!

**Comment by ps48** [comment]
@Cpruce I just checked the map_fn usage in your fork, it looks good to me. We don't need to add while_loop unnecessarily. It was just meant to be a replacement. I have done a few changes and sent a PR please check.

**Comment by Cpruce** [comment]
@ps48 thanks and your changes seem good to me! Just merged your commit into the branch. I'll start logical debugging tonight :)

**Comment by ps48** [comment]
@Cpruce do let me know if you face any issues!

**Comment by Cpruce** [comment]
@ps48 thanks again and will do! :D

**Comment by waleedka** [comment]
@Cpruce @ps48 Thanks for the contribution! Great progress.

The trick with unique_with_counts() is neat. But I think we can get set_intersection to work by adding a dimension at axis 0 and then picking the values of the output sparse tensor, as such:
```
# change this:
keep = tf.sets.set_intersection(keep, nms_keep)
# to this:
keep = tf.sets.set_intersection(tf.expand_dims(keep, 0), tf.expand_dims(nms_keep, 0))[1]
```

I didn't test it on the main code, but I did this simple test to verify.
```
with tf.Graph().as_default():
    with tf.Session() as session:
        a = np.array([1, 2, 3, 4, 5, 6], dtype=np.int32)
        b = np.array([1, 3, 5, 7, 9], dtype=np.int32)
        c = tf.sets.set_intersection(np.expand_dims(a, axis=0), 
                                     np.expand_dims(b, axis=0))
        r = session.run(c)
        print(r[1]) 

# output
[1 3 5]
```

And I noticed a bug in calling tf.where(). The return value of tf.where() is 2D tensor, unlike np.where() which returns a 1D result. This test demonstrates the issue:
```
with tf.Graph().as_default():
    with tf.Session() as session:
        a = np.array([1, 2, 3, 4, 5, 4, 3, 2, 1], dtype=np.int32)
        b = tf.where(a == 3)[0]
        c = tf.where(a == 3)[:, 0]
        r = session.run([b, c])
        print(r[0])
        print(r[1])

# Output
[2]      <-- wrong
[2 6]    <-- correct
```

**Comment by Cpruce** [comment]
Thanks @waleedka @ps48 for helping! 

@waleedka I have a few questions:

1. I faced a problem with `nms_keep` being sparse. Should I convert to dense here?
```python 
keep = tf.sets.set_intersection(
             tf.expand_dims(keep, 0),
            tf.expand_dims(tf.sparse_to_dense(nms_keep), 0))[1]
```

2. Could you please give a quick commentary on each use of `tf.where` in `refine_detections_graph`? I had no problem extracting the 0-dimension for the first two `tf.where`s. However, the one in the `nms_keep_map` function throws an error.


To replicate how I've been testing, you just need to put `pycocotools` and an old model (incorrect but helps) in the same directory and run `python3 coco.py evaluate --dataset=<path/to/coco> --model=coco`.

**Comment by waleedka** [comment]
I'm not sure why `nms_keep` is sparse. It shouldn't be, unless I'm missing something. I'd suggest tracing the code to see why you're getting a sparse tensor, and if you find out where that's happening and it's not a bug then I guess you'll need to convert to dense to run set_intersection.

And about the tf.where question. What's the error you're getting?

**Comment by Cpruce** [comment]
For the `tf.where` question,

```python
uniq_pre_nms_class_ids = tf.unique(pre_nms_class_ids)[0]

      nms_keep = []
      def nms_keep_map(class_id):
         #class_id = tf.expand_dims(class_id, -1)
          print('pre_nms_class_ids.shape', pre_nms_class_ids.shape)
          print('class_id', class_id.shape)
         ixs = tf.where(pre_nms_class_ids == class_id)

         print('ixs {}'.format(ixs.shape))
          # Apply NMS
          class_keep = tf.image.non_max_suppression(
                  tf.to_float(tf.gather(pre_nms_rois,ixs)),
                  tf.gather(pre_nms_scores, ixs),
                 max_output_size=ixs.shape[0],
                  iou_threshold=config.DETECTION_NMS_THRESHOLD)

          # Map indicies
          return tf.gather(keep, tf.gather(ixs, class_keep))

      print('uniq_pre_nms_class_ids: {}'.format(uniq_pre_nms_class_ids.shape))
      nms_keep = tf.to_int64(tf.unique(tf.concat(
          tf.map_fn(nms_keep_map, uniq_pre_nms_class_ids), axis=0))[0])
```

```ValueError: Tried to convert 'max_output_size' to a tensor and failed. Error: Cannot convert an unknown Dimension to a Tensor: ?```

It seems that `ixs.shape[0]` is `?`. I have reshaped before but do not know the dimension size here. Got any recommendations?

Seems Ill have to do some more debugging as to why `nms_keep` is sparse.

File: model.py [hunk 1]
```diff
@@ -2487,19 +2619,11 @@ def parse_image_meta(meta):
     active_class_ids = meta[:, 8:]
     return image_id, image_shape, window, active_class_ids
 
-
-def parse_image_meta_graph(meta):
```
Comment (LandyGuo) [inline model.py]: this function is used in training phase,  should not be deleted

Comment (Cpruce) [reply model.py]: Definitely, I did not plan on committing this part. If you know the solution to the current problem, I can clean this up asap. You can view the current issue in the pull request's last comment.

Comment (ps48) [reply model.py]: Just checked it out! I got your point that the parse_image_meta_graph is being used in the training phase, my point was to remove the redundant function of parse_image_meta.  We'll do the clean up once our current errors are resolved. @LandyGuo Thanks for pointing it out 👍

Comment (LandyGuo) [reply model.py]: Your work is very impressive !  I have made some changes base on your implementation, and now everything seems fine. 

please refer to: https://github.com/LandyGuo/Mask_RCNN/blob/master/debug_detection_graph.py

Comment (Cpruce) [reply model.py]: Awesome job @LandyGuo 👍 Do you mind submitting a pull request so we can merge into the original pull request and get the reviews/testing underway?

Comment (ericj974) [reply model.py]: @LandyGuo Thanks for this ! I tried / tested your implementation on my side and it seems to work fine (final detections seem identical). No problem when exporting and then evaluating the frozen graph.

For consistency, I would just replace `scores_select_size = class_ids.shape[0]` by `scores_select_size = tf.shape(probs)[0]`

Comment (ps48) [reply model.py]: @ericj974  Thanks for the review! Great to hear it works well end to end. @LandyGuo Thanks for the awesome contribution. @Cpruce  I'll consolidate all the changes and send a PR to tonight so that we can merge this as soon as possible.

Comment (Cpruce) [reply model.py]: Sounds good. Can you merge the incremental changes into mine? I was going to do it if Landy didn't want to but it seems like Landy fell off the face of the Earth... he never responded to me reaching out about him claiming credit for his contribution.

**Comment by Cpruce** [comment]
@waleedka Hi Waleed, could you please review the branch now? The Coco validation results match master. Shoutout for the contributions from @ps48 and @LandyGuo !

check_results

**Comment by waleedka** [comment]
Great progress! I'm reviewing this now and will update when I'm done.

I noticed, though, that you checked in a big change in the `train_shapes` notebook. Can you please revert that? I'd rather not check-in irrelevant changes like those generated when you run a notebook and get different images. If there is a change in the code in that notebook, then check-in the affected lines only.

**Comment by waleedka** [comment]
This PR is finally merged! Thank you, everyone, for the great work. 

Other than a bit of cleaning up, the main thing I changed was replacing `tf.while_loop` with `tf.map_fn`. In loop version, each step depends on the result from the previous step to concatenate with it, which prevents TF from running steps in parallel. I changed so that each step in the loop is independent.

**Comment by ps48** [comment]
@waleedka thanks for the support 👍
